Implement my domains screen

mkevins opened the pull request:
Fixes #19072 

### Description

This PR implements the "My Domains" screen, with an entry point on the "Me" tab of the main activity.

To test:

Pre-requisites: Please test this screen with an account with active domains.

1. Enable the Domain Management feature flag (in debug settings)
2. Open the My domains screen: Me -> Domains -> My Domains

The initial "ghost" screen should load while the domains list is fetched. It should then display the list of domains as pictured below (in various modes):

| State | Light | Dark | RTL |
|-|-|-|-|
| `Initial` | ![image](https://github.com/wordpress-mobile/WordPress-Android/assets/8507675/c523f468-acc9-4907-abca-ce8c8d6059b2) | ![image](https://github.com/wordpress-mobile/WordPress-Android/assets/8507675/29e89565-7c16-4471-95fd-dd1a0959a1e6) | ![image](https://github.com/wordpress-mobile/WordPress-Android/assets/8507675/03d6275f-2cc1-4cc1-9a7d-4868c72fcc55) |
| `Loaded` | ![my-domains-light](https://github.com/wordpress-mobile/WordPress-Android/assets/8507675/780f12a3-5b11-452b-8aad-b1d482653200) | ![my-domains-dark](https://github.com/wordpress-mobile/WordPress-Android/assets/8507675/c8b681d6-ea6a-40d2-beb8-a51fbb0f57cb) | ![image](https://github.com/wordpress-mobile/WordPress-Android/assets/8507675/5ac3957c-573f-4c75-9ab9-0d4eb6c347dd) |


## Regression Notes
1. Potential unintended areas of impact
N/A

4. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

5. What automated tests I added (or what prevented me from doing so)
Unit tests will be added later, as the view model logic is wired up to other flows.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

mkevins committed:
Add tracking event for "My domains" screen

mkevins committed:
Add viewmodel for domain management

mkevins committed:
Add my domains screen composable

mkevins committed:
Add domain management activity

mkevins committed:
Add entry-point on me screen for domain management activity

Note: this is currently using the ActivityLauncher, but may be changed
later.

mkevins committed:
Update viewmodel to fetch from the all-domains endpoint

mkevins committed:
Display populated lists of domains on my domains screen

antonis commented on WordPress/src/main/java/org/wordpress/android/ui/domains/management/MyDomainsScreen.kt:
```diff
@@ -0,0 +1,105 @@
+                    domain = it.domain,
+                    title = null,
+                    status = it.domainStatus,
+                    expiry = it.expiry?.time?.let { epoch ->
```
I got an error in the `LocalDate` conversion and tried out a [different conversion from here](https://github.com/wordpress-mobile/WordPress-Android/blob/57a414e6b95a5216cdbb32909208daa1a384b921/WordPress/src/main/java/org/wordpress/android/ui/jetpackoverlay/JetpackFeatureRemovalBrandingUtil.kt#L98) that worked for me.
```
FATAL EXCEPTION: main
Process: com.jetpack.android.beta, PID: 18535
j$.time.DateTimeException: Unable to obtain LocalDate from TemporalAccessor: 2024-03-24T00:00:00Z of type j$.time.Instant
```

For reference my test data are the following:
```
{
  "domains": [
    {
      "domain": "antonis.me",
      "blog_id": 58348,
      "blog_name": "antonis.log()",
      "type": "mapping",
      "is_domain_only_site": false,
      "is_wpcom_staging_domain": false,
      "has_registration": false,
      "registration_date": "2009-03-26T21:20:53+00:00",
      "expiry": "2024-03-24T00:00:00+00:00",
      "wpcom_domain": false,
      "current_user_is_owner": true,
      "site_slug": "antonis.me",
      "domain_status": {
        "status": "Active",
        "status_type": "success"
      }
    },
    {
      "domain": "antonistest1.blog",
      "blog_id": 217792462,
      "blog_name": "Site Title",
      "type": "mapping",
      "is_domain_only_site": false,
      "is_wpcom_staging_domain": false,
      "has_registration": false,
      "registration_date": "1970-01-01T00:00:00+00:00",
      "expiry": false,
      "wpcom_domain": false,
      "current_user_is_owner": false,
      "site_slug": "antonistest1.wordpress.com",
      "domain_status": {
        "status": "Expiring soon",
        "status_type": "error",
        "status_weight": 1000
      }
    }
  ]
}
```

The below suggestion works for me along the import change https://github.com/wordpress-mobile/WordPress-Android/pull/19366/files#r1358288149
```suggestion
                    expiry = it.expiry?.toInstant()?.atZone(ZoneId.systemDefault())?.toLocalDate()
```

mkevins replied:
Thanks for the suggestion Antonis! I updated this with https://github.com/wordpress-mobile/WordPress-Android/pull/19366/commits/59df94b93802acf371f2f17ac458bd7aa540e634, otherwise, the suggestions on GitHub would have created two commits.

antonis commented on WordPress/src/main/java/org/wordpress/android/ui/domains/management/MyDomainsScreen.kt:
```diff
@@ -0,0 +1,105 @@
+import org.wordpress.android.fluxc.network.rest.wpcom.site.AllDomainsDomain
+import org.wordpress.android.ui.compose.components.MainTopAppBar
+import org.wordpress.android.ui.compose.components.NavigationIcons
+import java.time.Instant
+import java.time.LocalDate
```
```suggestion
import java.time.ZoneId
```

antonis commented on WordPress/src/main/java/org/wordpress/android/ui/domains/management/MyDomainsScreen.kt:
```diff
@@ -0,0 +1,105 @@
+            DomainListCard(
+                uiState = DomainCardUiState(
+                    domain = it.domain,
+                    title = null,
```
I've tested adding the `blogName` field here to match the data I see on the web.

```suggestion
                    title = it.blogName,
```
|Web|Mobile|
|---|---|
|![Screenshot 2023-10-13 at 4 21 42 PM](https://github.com/wordpress-mobile/WordPress-Android/assets/304044/dde3c512-0607-4772-82b5-ab32610f6169)|![Screenshot_20231013_165504](https://github.com/wordpress-mobile/WordPress-Android/assets/304044/0e095615-b586-4dde-8b2e-40eb25dc3de2)|

mkevins replied:
Nice! Thanks Antonis!

antonis reviewed (commented):
Awesome work @mkevins ⭐ 
I made an initial pass of the code and tested on a Pixel 5 (Android 14). Things worked pretty well for me considering that the PR is still draft  🎉

I've added a couple of comments/suggestions I tested to make the screen work with my test user data.
I've also started an internal discussion related to the missing/loading value of my second domain (ref p1697206669077769-slack-C05NS0YV7HS)

I also noticed two minor UI issues in my tests. 

One has to do with the header background in dark mode.
|Dark mode|Default|
|---|---|
|![Screenshot_20231013_170033](https://github.com/wordpress-mobile/WordPress-Android/assets/304044/b3f130a2-4e25-48ec-a19a-356572d09ce3)|![Screenshot_20231013_165949](https://github.com/wordpress-mobile/WordPress-Android/assets/304044/aa8f964d-24ae-40f7-8d3a-4372928c92fe)|

The second is related to the scroll state (reproduced only in landscape with my data). I would suggest adding a shadow when scrolling below the search field or scroll the whole screen with the search field. cc @paulforgione for design feeback 🙏 

|Landscape|Landscape scrolled|
|---|---|
|![Screenshot_20231013_165957](https://github.com/wordpress-mobile/WordPress-Android/assets/304044/8197651e-2cc6-4f53-81c8-ac87d2f6d809)|![Screenshot_20231013_170007](https://github.com/wordpress-mobile/WordPress-Android/assets/304044/802e92d2-5918-4e63-b3fa-c4a29ec6ed19)|

paulforgione commented:
Looking great! I think a shadow would work perfectly fine in this case, or a light shade of grey to provide some contrast, assuming one is specified ([as in the M3 guidelines](https://m3.material.io/components/top-app-bar/guidelines#4eab4f50-4a3e-4189-bce2-a46514cde1da)).

mkevins committed:
Merge branch 'issue/19074-domains-list-card-ui' into issue/19072-my-domains-screen

mkevins committed:
Use `blogName` from AllDomainsDomain for `title`

Co-authored-by: Antonis Lilis <antonis.lilis@automattic.com>

mkevins committed:
Add my domains screen preview for night mode

mkevins committed:
Fix Date to LocalDate conversion

Thanks Antonis for the suggestion!

mkevins committed:
Allow overriding colors and elevation in MainTopAppBar

This is needed for compatibility with Material 3 theming. In the future,
a Material 3 MainTopAppBar can be used, but that API is experimental for
now, and subject to change.

mkevins committed:
Fix top bar dark mode issue

mkevins committed:
Add elevation to search input when domain list is scrolled

mkevins commented:
Thank you for reviewing and testing this Antonis, and thank you for your feedback Paul!

> I also noticed two minor UI issues in my tests.

Good catches!

> One has to do with the header background in dark mode.

I've fixed this with https://github.com/wordpress-mobile/WordPress-Android/pull/19366/commits/3df164f986340fdabfec616332973c206f2c8a0a

> The second is related to the scroll state (reproduced only in landscape with my data). I would suggest adding a shadow when scrolling below the search field

I've addressed this with https://github.com/wordpress-mobile/WordPress-Android/pull/19366/commits/2fb2ddfd749a5ee5746aeef19558886203cc0e42

Note: There are still a few things I need to clean up in this PR before I make it ready for another review: e.g. string extraction, and adding a "+" button to the top app bar.

mkevins committed:
Add plus icon to domains list top bar

mkevins committed:
Fix back icon for RTL layout direction

This fixes the back icon in the MainTopAppBar. Hat tip to Artyom for
suggesting the fix.

mkevins committed:
Merge branch 'issue/19074-domains-list-card-ui' into issue/19072-my-domains-screen

mkevins committed:
Adapt domain management screen ui to refactored data classes

mkevins committed:
Adjust content padding in domain managment list

mkevins committed:
Extract string resources for my domains screen

mkevins committed:
Merge branch 'trunk' into issue/19072-my-domains-screen

mkevins committed:
Hide chevron for initial ("ghost") state

mkevins committed:
Add content description to domain details chevron for accessibility

mkevins commented:
This PR is ready for another pass.

antonis commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -2709,6 +2709,20 @@
+    <string name="purchase_domain_screen_existing_domain_card_title">Existing Wordpress.com site</string>
+    <string name="purchase_domain_screen_existing_domain_card_description">Use with a site you already started.</string>
+    <string name="purchase_domain_screen_existing_domain_card_button">Choose Site</string>
+    <string name="purchase_domain_screen_discount_notice">*A free domain for one year is included with all paid annual plans.</string>
```
Thank you for moving the two resource section together 👍

antonis reviewed (approved):
Great work @mkevins 👍 
Thank you for the changes 🙇 
I've retested the implementation on a Pixel 5 (Android 14) in different configurations and everything worked as expected.
The code code changes also LGTM 🎉 

|Portrait|Landscape|
|---|---|
|![day](https://github.com/wordpress-mobile/WordPress-Android/assets/304044/e276593c-33b3-4328-89f7-6cfc3d5d4aae)|![land](https://github.com/wordpress-mobile/WordPress-Android/assets/304044/cf266f2a-cc72-4c07-86bd-6d7d25b02130)|
|Large display|Large font|
|![screen](https://github.com/wordpress-mobile/WordPress-Android/assets/304044/b037e086-7797-49d4-8f6e-5b3471eb5def)|![font](https://github.com/wordpress-mobile/WordPress-Android/assets/304044/572df7e9-0450-406d-b261-d857584c4769)|

mkevins commented:
Thank you for reviewing, testing, and merging 😄
